File: two_pointers/trapped_water/trapped_water.java
class Solution {
  public int trap(int[] height) {
      int water = 0;
      int left = 0;
      int right = height.length - 1;
      int leftMax = height[left];
      int rightMax = height[right];

      while (left < right) {
          if (leftMax < rightMax) {
              left++;
              if (height[left] > leftMax) {
                  leftMax = height[left];
              }
              water += leftMax - height[left];
          } else {
              right--;
              if (height[right] > rightMax) {
                  rightMax = height[right];
              }
              water += rightMax - height[right];
          }
      }

      return water;
  }
}

File: two_pointers/trapped_water/trapped_water.py
class Solution:
    def trap(self, height):
        water = 0
        l = 0
        r = len(height) - 1

        lMax, rMax = height[l], height[r]
        while l < r:
            if lMax < rMax:
                l+=1
                lMax = max(lMax, height[l])
                water += lMax - height[l]
            else:
                r-=1
                rMax = max(rMax, height[r])
                water += rMax - height[r]
        return water
        
 